fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveUnionCursor.java

```diff
@@ -0,0 +1,264 @@
+        return CompletableFuture.completedFuture(RecordCursorResult.withNextValue(innerCursorResult.get(), continuation));
+    }
+
+    @Override
+    public void close() {
+        initialCursor.close();
+        recursiveCursor.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return recursiveCursor.isClosed(); // todo: improve.
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return executor;
+    }
+
+    @Override
+    public boolean accept(@Nonnull final RecordCursorVisitor visitor) {
+        if (visitor.visitEnter(this)) {
+            initialCursor.accept(visitor);
+            recursiveCursor.accept(visitor);
+        }
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    public static <T> RecursiveUnionCursor<T> from(@Nullable final byte[] unparsed,
+                                                   @Nonnull final Function<ByteString, RecordCursor<T>> initialCursorCreator,
+                                                   @Nonnull final Function<ByteString, RecordCursor<T>> recursiveCursorCreator,
+                                                   @Nonnull final Supplier<Boolean> isReadingFromInitialCursorSupplier,
+                                                   @Nonnull final Consumer<Boolean> wasInInitialStateConsumer,
+                                                   @Nonnull final Consumer<Boolean> wasReadingFromFirstTempTableConsumer,
+                                                   @Nonnull final Supplier<Boolean> recursiveStepCompletionCallback,
+                                                   @Nonnull Executor executor) {
+        if (unparsed == null) {
+            final var initialCursor = initialCursorCreator.apply(null);
+            final var recursiveCursor = recursiveCursorCreator.apply(null);
+            return new RecursiveUnionCursor<>(initialCursor, () -> recursiveCursorCreator.apply(null), recursiveCursor, executor, isReadingFromInitialCursorSupplier,
+                    recursiveStepCompletionCallback, true);
+        } else {
+            RecordCursorProto.RecursiveCursorContinuation proto;
+            try {
+                proto = RecordCursorProto.RecursiveCursorContinuation.parseFrom(unparsed);
+            } catch (InvalidProtocolBufferException ex) {
+                throw new RecordCoreException("invalid continuation", ex)
+                        .addLogInfo(LogMessageKeys.RAW_BYTES, ByteArrayUtil2.loggable(unparsed));
+            }
+            // resume the state.
+            wasInInitialStateConsumer.accept(proto.getIsInitialState());
+            wasReadingFromFirstTempTableConsumer.accept(proto.getIsReadingInitialCursor());
+            final var initialCursor = initialCursorCreator.apply(proto.getInitialCursorContinuation());
+            final var recursiveCursor = recursiveCursorCreator.apply(proto.getRecursiveCursorContinuation());
```

Comment:
Should this be conditional on the result of `getIsReadingInitialCursor`?

Reply:
Thanks, that's correct, the new implementation creates exactly _one_ cursor upon initialization, based on the initial state condition:

```java
                if (isInitialState) {
                    activeCursor = initialCursorCreator.apply(continuation.getActiveStateContinuation().toByteString(), overridenEvaluationContext);
                } else {
                    if (continuation.buffersAreFlipped()) {
                        overridenEvaluationContext = flipBuffers(baseContext, false);
                    }
                    activeCursor = recursiveCursorCreator.apply(continuation.getActiveStateContinuation().toByteString(), overridenEvaluationContext);
                }
```